mono/utils/mono-merp.c

```diff
@@ -0,0 +1,490 @@
+	while (true) {
+		for (int i = 0; BOOTSTRAP_UNKNOWN_SERVICE == kernErr && i < 5000; i++)
+			kernErr = bootstrap_look_up(bootstrap_port, serviceName, merp_port);
+
+		if (kernErr != BOOTSTRAP_UNKNOWN_SERVICE)
+			break;
+
+		if (config.log)
+			fprintf (stderr, "Merp: Service not registered with name %s, resetting counter after 10s sleep\n", serviceName);
+		sleep(10);
+	}
+
+	/*// FIXME error handling*/
+	g_assert (KERN_SUCCESS == kernErr);
+}
+
+static void
+mono_merp_send (GString *str, const char *serviceName)
+{
+	// Write struct to magic file location
+	// This registers our mach service so we can connect
+	// to the merp process
+	const char *home = g_get_home_dir ();
+	char *merpParamPath = g_strdup_printf ("%s/Library/Group Containers/UBF8T346G9.ms/MERP.Params.txt", home);
+	write_file (str, merpParamPath);
+	g_free (merpParamPath);
+
+	mach_port_t merpPort = MACH_PORT_NULL;
+
+	// Start merpGui application
+	// Assign to merpPort the port that merpGui opens
+	// Connecting to the service with name serviceName 
+	// that merpGUI starts
+	connect_to_merp (serviceName, &merpPort);
+
+	// Send this mach task to MERP over the port
+	send_mach_message (&merpPort);
+
+	// After we resume from the suspend, we are done
+	// We're spinning down the process shortly.
+	// This thread exits, and the crashing thread's
+	// waitpid on this thread ends.
+	return;
+}
+
+static void
+mono_init_merp (const char *serviceName, const char *signal, pid_t crashed_pid, intptr_t thread_pointer, MERPStruct *merp)
+{
+	g_assert (mono_merp_enabled ());
+
+	merp->pidArg = crashed_pid;
+	merp->archArg = get_merp_arch ();
+	merp->capabilitiesArg = 0x0;
+
+	merp->threadArg = thread_pointer;
+
+	// FIXME: time the runtime?
+	merp->timeArg = 0x0;
+
+	merp->exceptionArg = parse_exception_type (signal);
+
+	// If these aren't set, icall wasn't made
+	// don't do merp? / don't set the variable to use merp;
+	merp->bundleIDArg = config.appBundleID;
+	merp->signatureArg = config.appSignature;
+	merp->versionArg = config.appVersion;
+
+	// FIXME: Do we want these?
+	merp->devRegionArg = NULL;
+	merp->uiLidArg = NULL;
+
+	merp->serviceNameArg = serviceName;
+
+	// FIXME: Do we want these?
+	merp->appLoggerName = NULL;
+	merp->appLogSessionUUID = NULL;
+
+	// Should be set for our usage as per document
+	merp->isOfficeApplication = TRUE;
+	merp->isObjCException = FALSE;
+
+	mono_arch_memory_info (&merp->memRes, &merp->memVirt);
+
+	// No sessions right now
+	merp->treSessionID = NULL;
+	merp->exceptionCodeArg = NULL;
+	merp->exceptionAddressArg = NULL;
+
+	// Not certain? Maybe expose to config options
+	merp->isAppRegisteredWithMAU = TRUE;
+}
+
+void
+mono_merp_invoke (pid_t crashed_pid, intptr_t thread_pointer, const char *signal)
+{
+	// This unique service name is used to communicate with merp over mach service ports
+	char *serviceName = g_strdup_printf ("com.mono.merp.%.8x", crashed_pid);
```

Comment:
shouldn't this be freed?

Reply:
We're calling exit(1) as soon as this function returns, but sure, I can free it.

Reply:
Fixed at https://github.com/mono/mono/pull/7451/files#diff-11deec7182d4d43c390fa69013d4213aR462